rfcs/20230621-tf-api-deprecation.md

```diff
@@ -0,0 +1,101 @@
+# TensorFlow API Deprecation Charter
+
+| Status        | Proposed       |
+:-------------- |:---------------------------------------------------- |
+| **RFC #**     | [450](https://github.com/tensorflow/community/pull/450) |
+| **Author(s)** | Ayush Jain (ayh@google.com)                          |
+| **Sponsor**   | Rohan Jain (rohanj@google.com)                       |
+| **Updated**   | 2023-06-21                                           |
+
+## Objective
+
+This proposal outlines an explicit policy for long-term code health by adding a process and mechanism for API / op deprecation and removal, specific to TensorFlow.
+
+Because of TensorFlow’s position in both the Google-internal and third-party ML development worlds, we take special care when deprecating or removing public APIs. 
+
+Note: As a primarily process RFC, there are a few sections of the RFC template not included in this draft proposal.
+
+## Motivation
+
+The TensorFlow code base has grown continuously over many years, increasing in complexity to support multiple hardware backends, supporting eager and graph modes, and more. TensorFlow has followed the Semantic Versioning policy, with [strict backward compatibility guarantees](https://www.tensorflow.org/guide/versions) that only allows for backward incompatible changes to the public Python API in a major version release. 
+
+TensorFlow had only one major version revision, and this strict adherence to backward compatibility has come with a high maintenance burden as well as a complicated and occasionally redundant API surface for users. By establishing a mechanism to clean up unnecessary and/or unused behaviors, TensorFlow's users, contributors, and developers alike can build with greater simplicity and speed.
+
+## Proposal
+
+### Eligibility
+
+An API is eligible for deprecation whenever there exists a better way to fulfill its purpose, making alternatives obsolete. “Better” and “obsolete” are terms subject to maintainer discretion - the API may either be part of an unsupported and since-replaced suite, exhibit inferior performance, not comply with Function-based execution, or otherwise. As defined, an eligible API must then have a functional, non-experimental equivalent available that users can be directed upon deprecation. 
+
+If there are fewer than fewer than 500 dependent non-TF-fork repositories on GitHub, said API is deemed suitable for public deprecation. If it does not meet this requirement, it is ineligible and the process stops here. If usage diminishes over time, it can be reevaluated at a later date.
+
+Note that while we use API as a blanket term throughout this document, this policy also explicitly covers TensorFlow ops that have fallen out of favor. For an op to be eligible for deprecation, it must be expressible through a graph rewrite using other existing ops.
```

Comment:
> must be expressible through a graph rewrite using other existing ops

Must we provide the graph rewrite for the deprecated op until the point at which the op is removed?

Reply:
You would need the rewrite specifically after removal, no? So any minimal existing usage doesn't break irreparably?

Reply:
I assumed it wouldn't load successfully if the op no longer existed (prior to rewrite).  But I'm not sure of that.  I was thinking graph rewrite after deprecation but before removal so that if you saved the model (either SavedModel or checkpoint), the saved version would contain the new op (rather than the deprecated one).

But what I mean to ask was whether we need a graph rewrite pass as part of removing an op.  Later in the doc it talks more about adding documentation about how to replace deprecated calls.  There's no mention of _requiring_ a graph rewrite pass to update models.

Reply:
Ah, that's fair. We'd want serialized models to use the new/replacement op expression.

I think we would need a rewrite as part of initial deprecation. cc @rohan100jain. I'll add language about requiring as such, good point.